airflow/models/dag_edge.py

```diff
@@ -0,0 +1,60 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from sqlalchemy import Column, String, Index, Integer
+
+from airflow.utils.db import create_session
+
+from airflow.models.base import Base
+
+
+class DagEdge(Base):
+    """
+    Dags can originate from different places (user repos, master repo, ...)
+    and also get executed in different places (different executors). This
+    object represents a version of a DAG and becomes a source of truth for
+    a BackfillJob execution. A pickle is a native python serialized object,
+    and in this case gets stored in the database for the duration of the job.
+
+    The executors pick up the DagPickle id and read the dag definition from
+    the database.
+    """
+
+    __tablename__ = "dag_edge"
+
+    dag_id = Column(String(250), primary_key=True)
+    graph_id = Column(Integer, primary_key=True)
+    from_task = Column(String(250), primary_key=True)
+    to_task = Column(String(250), primary_key=True)
+
+    __table_args__ = (Index('idx_dag_edge', dag_id, graph_id, from_task, to_task, unique=True),)
```

Comment:
+1 on Ash previous command. Given all the keys are all primary keys, why we need to build another index?

Reply:
I think this depends a bit on the database implementation. I do think that some databases makes the index specific for 4 columns and therefore not compatible for search just 2 columns, this problem depends on the order of the columns. I think is does not harm to keep 2 indexes next to each other.